### PR TITLE
Document connecting Node-RED

### PR DESCRIPTION
**Pull request by nejraselimovic**
#### Summary
References #2510.  

#### Changes
Added documentation for connecting Node-RED mqtt input node to MQTT Server (that is exposed by Application Server on locally deployed TTS) and adding debug nodes for the purpose of listening published messages.

#### Testing
Manually by listening messages from the locally connected device.

#### Checklist
- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [x] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [ ] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+First, you need to run the Node-RED by using the following command in terminal:
+```bash
+$ node-red 
+```
+```bash
+# Tip: Node-RED runs locally on port 1880 by default, so make sure to allow the service to run on this port with:
+
+$ sudo ufw allow 1880
+```
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:22]: Use empty lines between blocks. See also the documentation requirements: https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/DEVELOPMENT.md#documentation-site

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
+---
+title: "Connecting to Node-RED"
+description: ""
+---
+
+[Node-RED](https://nodered.org/) is free, JavaScript-based development tool for visual programming, developed to ease the process of wiring together hardware devices, APIs and online services.
+
+The MQTT server that is exposed by the Application Server can be connected to Node-RED. This kind of integration allows setting up a Node-RED flow that listens to events and messages sent by the end devices. Also, it is possible to send messages to the end devices.
+
+This guide contains the instructions for setting up these flows, but first make sure to properly [install the Node-RED](https://nodered.org/docs/getting-started/local).
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:10]: Add a "summary separator" here somewhere. See also the documentation guidelines: https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/DEVELOPMENT.md#documentation-site

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+## Setup
+
+First, you need to run the Node-RED by using the following command in terminal:
+```bash
+$ node-red 
+```
+```bash
+# Tip: Node-RED runs locally on port 1880 by default, so make sure to allow the service to run on this port with:
+
+$ sudo ufw allow 1880
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:21]: You only need this if you want to expose Node-RED to the Internet, which is out of scope for our guide. For our guide it's sufficient to show how to connect to Node-RED instance running locally only.

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+```
+
+>Note: you have to [secure Node-RED](https://nodered.org/docs/user-guide/runtime/securing-node-red) if you want to connect it to MQTT Server over TLS, since its editor is not secured by the default.
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:24]: This isn't actually required to connect over TLS, right? Of course it's good practice to secure a Node-RED deployment, but that's out of scope for our guide about connecting it to The Things Stack.

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+{{< figure src="nodered_dashboard.png" alt="Node-RED dashboard" >}}
+
+On the left side, you can see various types of nodes (they can also be found in the [Node-RED library](https://flows.nodered.org/)) that can be used in order to build flows, although in this guide only a few simple inbuilt nodes are used. 
+
+The Console provides the information needed for completing this integration, under **MQTT** submenu from **Integrations** menu within desired application.
+
+{{< figure src="console_info.png" alt="MQTT Server connection information" >}}
+
+In this example, since TTS is running locally, the inbuilt MQTT Server is deployed on port 1883 for insecure connection and on port 8883 for the TLS-secured one.
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:36]: Don't write `TTS`. That's not the name of our product. You can use the `{{% tts %}}` shortcode. Again, see the guidelines: https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/DEVELOPMENT.md#documentation-site

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+## Receive events and messages
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:38]: Use Title Case for headings. Again, see the guidelines: https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/DEVELOPMENT.md#documentation-site

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+This section follows the process of setting up a flow which subscribes and listens to the events and messages that are being published by the MQTT Server. 
+
+First, find the `mqtt in` node in the library section and place it on the flow dashboard. This node connects to the MQTT broker and subscribes to messages from the specified topic. Double-click on the node to configure its properties.
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:42]: I think we should also use `**Strong**` when referring to nodes, similar to how we style text that refers to buttons.

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+{{< figure src="node_connection.png" alt="COnfiguring MQTT Server connection information" >}}
+
+In the **Connection** tab, under **Server**, you should provide MQTT server's URL - if it runs locally, then type **localhost**. The **Port** value depends on whether the connection is to be set up as a secure one or not. In this example, we are using TLS, so the **Port** value is set to 8883. In this case you should also check the **Enable secure (SSL/TLS) connection** box.
+
+{{< figure src="node_security.png" alt="Configuring MQTT Server credentials" >}}
+
+In the **Security** tab, enter the **Username** and **Password** according to values in the Console.
+
+Under **Topic**, a topic to listen to can be specified, but for the testing purposes set the value to `#` (all topics). The **QoS** value can be arbitrarily chosen, as well as the **Output**. In this example, we want to see each message as `a parsed JSON object`.
```
Comment (htdvisser) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:56]: 1. Shouldn't we recommend subscribing to specific topics instead of using "all"?
2. QoS is not arbitrary. Let's leave it at 0.

**Comment by KrishnaIyer** [comment]
@nejraselimovic: Please reduce the number of reviewers 
(unless it's something the entire team should know about, like process/api/tooling changes).
You can check the [CODEOWNERS](https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/CODEOWNERS) file to see who owns a particular function and ask them to review. 
When in doubt, you can set @laurensslats who can then assign it.. 👍

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md (hunk 4 above)
Comment (nejraselimovic) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:24]: I actually did have problems with connecting Node-RED's MQTT input node to MQTT server over port 8883, until I followed these steps for running Node-RED over https, so I thought those were required, but I guess something else could've been the issue there

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md (hunk 8 above)
Comment (nejraselimovic) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md:56]: 1. Subscribing to specific topics is already explained a step back in the docs, in https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/doc/content/integrations/mqtt/_index.md, and there is also a usage of '#' for topic name (I guess to make it easier to follow). Do you still think that I should change this?
2. I guess I meant arbitrarily chosen from the listed options according to one's needs, but I will rephrase this as you said.

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 9]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
->Note: you have to [secure Node-RED](https://nodered.org/docs/user-guide/runtime/securing-node-red) if you want to connect it to MQTT Server over TLS, since its editor is not secured by the default.
 
 If Node-RED is sucessfully launched, when navigating to `https://localhost:1880` in your web browser you should see something like:
 
 {{< figure src="nodered_dashboard.png" alt="Node-RED dashboard" >}}
 
 On the left side, you can see various types of nodes (they can also be found in the [Node-RED library](https://flows.nodered.org/)) that can be used in order to build flows, although in this guide only a few simple inbuilt nodes are used. 
 
-The Console provides the information needed for completing this integration, under **MQTT** submenu from **Integrations** menu within desired application.
+The in the {{% tts %}} Console provides the information needed for completing this integration, under **MQTT** submenu from **Integrations** menu within desired application.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: Remove "The in the"; this will get rendered as "The Things Stack Console provides ..." which is correct

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 10]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
 
 {{< figure src="console_info.png" alt="MQTT Server connection information" >}}
 
-In this example, since TTS is running locally, the inbuilt MQTT Server is deployed on port 1883 for insecure connection and on port 8883 for the TLS-secured one.
+In this example, since {{% tts %}} is running locally, the inbuilt MQTT Server is deployed on port 1883 for insecure connection and on port 8883 for the TLS-secured one.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: - built-in
- deployed => configured by default
- insecure connections (plural)

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 11]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
 
-## Receive events and messages
+## Receive Events and Messages
 
 This section follows the process of setting up a flow which subscribes and listens to the events and messages that are being published by the MQTT Server. 
 
-First, find the `mqtt in` node in the library section and place it on the flow dashboard. This node connects to the MQTT broker and subscribes to messages from the specified topic. Double-click on the node to configure its properties.
+First, find the **mqtt in** node in the library section and place it on the flow dashboard. This node connects to the MQTT broker and subscribes to messages from the specified topic. Double-click on the node to configure its properties.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: Good to mention that this is in the **input** category

In my Node-RED (0.20.7) this is called **mqtt** node. The output node is also **mqtt**, but in the **output** category

Comment (johanstokking) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: Just updated to Node-RED 1.0.6 and now it's in the **network** category, indeed as **mqtt {in/out}**. I think it would be helpful to mention the Node-RED version on top of this guide

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 12]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
+{{< figure src="mqtt_in_node_properties.png" alt="mqtt in node properties" >}}
 
 In the **Server** dropdown menu, select **Add new mqtt-broker** and click on the button besides to edit it. 
 
-{{< figure src="node_connection.png" alt="COnfiguring MQTT Server connection information" >}}
+{{< figure src="mqtt_in_node_connection.png" alt="COnfiguring MQTT Server connection information" >}}
 
-In the **Connection** tab, under **Server**, you should provide MQTT server's URL - if it runs locally, then type **localhost**. The **Port** value depends on whether the connection is to be set up as a secure one or not. In this example, we are using TLS, so the **Port** value is set to 8883. In this case you should also check the **Enable secure (SSL/TLS) connection** box.
+In the **Connection** tab, under **Server**, you should provide MQTT Server's URL - if it runs locally, then type **localhost**. The **Port** value depends on whether the connection is to be set up as a secure one or not. In this example, TLS is being used, so the **Port** value is set to 8883. In this case you should also check the **Enable secure (SSL/TLS) connection** box.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: >MQTT Server's URL - if it runs locally, then type **localhost**.

- It's not a URL; it's an address
- They can find the address in MQTT under Integrations. That is the only place we should refer to; do not mention "if it runs locally" in this guide; if the server runs locally, the address shown in the Console is already `localhost`

Screenshot 2020-05-29 at 09 10 32

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 13]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
 
-{{< figure src="node_security.png" alt="Configuring MQTT Server credentials" >}}
+{{< figure src="mqtt_in_node_security.png" alt="Configuring MQTT Server credentials" >}}
 
-In the **Security** tab, enter the **Username** and **Password** according to values in the Console.
+In the **Security** tab, enter the **Username** and **Password** according to values in the in the {{% tts %}} Console.
 
-Under **Topic**, a topic to listen to can be specified, but for the testing purposes set the value to `#` (all topics). The **QoS** value can be arbitrarily chosen, as well as the **Output**. In this example, we want to see each message as `a parsed JSON object`. 
+Under **Topic**, a topic to listen to can be specified, but for the testing purposes set the value to `#` (all topics). The **QoS** value can be selected from the listed options, as well as the **Output**. In this example, we want to see each message as `a parsed JSON object`.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: **a parsed JSON object**

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 14]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
-{{< figure src="node_security.png" alt="Configuring MQTT Server credentials" >}}
+{{< figure src="mqtt_in_node_security.png" alt="Configuring MQTT Server credentials" >}}
 
-In the **Security** tab, enter the **Username** and **Password** according to values in the Console.
+In the **Security** tab, enter the **Username** and **Password** according to values in the in the {{% tts %}} Console.
 
-Under **Topic**, a topic to listen to can be specified, but for the testing purposes set the value to `#` (all topics). The **QoS** value can be arbitrarily chosen, as well as the **Output**. In this example, we want to see each message as `a parsed JSON object`. 
+Under **Topic**, a topic to listen to can be specified, but for the testing purposes set the value to `#` (all topics). The **QoS** value can be selected from the listed options, as well as the **Output**. In this example, we want to see each message as `a parsed JSON object`. 
 
 >Note: instead of changing **Output** parameter to `a parsed JSON object`, you can also use a `JSON` node or the `function` node with proper parsing function.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: This is not relevant

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md (hunk 13 above)
Comment (johanstokking) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: Also add a reference to the MQTT topic structure; `/integrations/mqtt`

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 15]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
 
-Next, add two `debug` nodes to the flow and connect both of them to the `mqtt in` node. One debug node will listen to the events, while the other will listen to messages coming from the devices. To make this happen, set the **Output** parameter for the first debug node configuration to `complete msg object` value, and to `msg.payload` for the other. 
+Next, add two **debug** nodes to the flow and connect both of them to the **mqtt in** node. One debug node will listen to the events, while the other will listen to messages coming from the devices. To make this happen, set the **Output** parameter for the first debug node configuration to `complete msg object` value, and to `msg.payload` for the other.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: >will listen to messages coming from the devices

This is not entirely true if you subscribe to `#` as you get a lot of other events which are not messages from end devices; those would only be `v3/{application id}/devices/{device id}/up`

File: doc/content/integrations/mqtt/Connecting to Node-RED/_index.md [hunk 16]
```diff
@@ -15,50 +17,69 @@ First, you need to run the Node-RED by using the following command in terminal:
...
+{
+  "downlinks": [{
+    "f_port": 15,
+    "frm_payload": "vu8=",
+    "priority": "NORMAL"
+  }]
+}
+```
+
+in order to schedule a downlink message with hexadecimal payload `BE EF`. State the period between the automatic injections if you want them or choose `None` for **Repeat** if you do not.
```
Comment (johanstokking) [inline doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: Inconsistent use of `code block` vs **literal** to be found in the screen vs "value" to be entered

Comment (nejraselimovic) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: Could you please clarify this?

Comment (johanstokking) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: - Use `code blocks` for code, file names, configuration keys, environment variables, etc, all the technical identifiers
- Use **strong** for text that you can find literally in the user interface, like **Gateways** or **User name** in the login form
- Use "quotes" for input

Here, you say `None`, while that should be "None"

I'm also open to suggestions for changing this, as long as we are consistent. cc @benolayinka

Comment (nejraselimovic) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: The word 'none' here is actually not a value to be entered (although it looks like that because of the small case), it's being chosen from the dropdown list. I guess that means that it should be **none**, to be consistent with **a parsed JSON object**

Comment (benolayinka) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: > * Use `code blocks` for code, file names, configuration keys, environment variables, etc, all the technical identifiers
> * Use **strong** for text that you can find literally in the user interface, like **Gateways** or **User name** in the login form
> * Use "quotes" for input
> 
> Here, you say `None`, while that should be "None"
> 
> I'm also open to suggestions for changing this, as long as we are consistent. cc @benolayinka

It looks good the way it is but is it documented anywhere?

I'll push a 'copy me' draft template to the docs with a skeleton of a docs page, and examples of these things. I can also add it to CONTRIBUTING.md if it isn't already somewhere else

Comment (htdvisser) [reply doc/content/integrations/mqtt/Connecting to Node-RED/_index.md]: Documentation guidelines are here: https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/DEVELOPMENT.md

**Comment by johanstokking** [comment]
Not seeing a new review request; please do that when ready

File: doc/content/integrations/node-red/setup/_index.md [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+## Setup
+
+Run Node-RED by using the following command in the terminal:
+
+```bash
+$ node-red 
+```
+
+With Node-RED running, navigate to `https://localhost:1880` (or the public  address of your Node-RED instance).
```
Comment (johanstokking) [inline doc/content/integrations/node-red/setup/_index.md:23]: Not https

File: doc/content/integrations/node-red/setup/_index.md [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+>Note: Node-RED v1.0.6 is current at time of writing and is used in this guide.
+
+## Setup
+
+Run Node-RED by using the following command in the terminal:
+
+```bash
+$ node-red 
+```
```
Comment (johanstokking) [inline doc/content/integrations/node-red/setup/_index.md:21]: We don't have to add anything to https://nodered.org/docs/getting-started/local; remove these instructions. There are better ways to run than this, as mentioned on that page and references above

File: doc/content/integrations/node-red/receive/_index.md [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+weight: 2
+---
+
+This section follows the process of setting up a flow which subscribes and listens to the events and messages that are being published by the MQTT Server. 
+
+## Configure MQTT In Node
+
+Find the **mqtt in** node in the library section and place it on the flow dashboard. This node connects to the MQTT broker and subscribes to messages from the specified topic. Double-click on the node to configure its properties.
+
+{{< figure src="mqtt_in_node_properties.png" alt="mqtt in node properties" >}}
```
Comment (johanstokking) [inline doc/content/integrations/node-red/receive/_index.md:13]: This screenshot already contains a configured `localhost:8883` server. Please remove, as it will be added in instructions below

Comment (johanstokking) [reply doc/content/integrations/node-red/receive/_index.md:13]: Also please put the screenshots under the instructions, instead of the screenshots above

File: doc/content/integrations/node-red/receive/_index.md [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+{{< figure src="mqtt_in_node_connection.png" alt="Configuring MQTT Server connection information" >}}
+
+In the **Connection** tab, under **Server**, provide the MQTT Server address from {{% tts %}} Console. The **Port** value depends on whether the connection is to be set up as a secure one or not. In this example, TLS is used, so the **Port** value is set to 8883. In this case you should also check the **Enable secure (SSL/TLS) connection** box.
+
+{{< figure src="mqtt_in_node_security.png" alt="Configuring MQTT Server credentials" >}}
+
+In the **Security** tab, enter the **Username** and **Password** according to the values in {{% tts %}} Console.
+
+Under **Topic**, a topic to listen to can be specified. For testing purposes, we will set the value to "#" (all topics). The **QoS** value can be selected from the listed options, as well as the **Output**. In this example, we want to see each message as **a parsed JSON object**. A full list of topics that you can subscribe to is mentioned in [MQTT Server]({{< ref "/integrations/mqtt" >}}) guide.
```
Comment (johanstokking) [inline doc/content/integrations/node-red/receive/_index.md:25]: Write `#` as code (not "#")

File: doc/content/integrations/node-red/receive/_index.md [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+Next, add two **debug** nodes to the flow and connect both of them to the **mqtt in** node. One debug node will listen to the events, while the other will listen to published messages. To make this happen, set the **Output** parameter for the first debug node configuration to **complete msg object** value, and to **msg.payload** for the other.
```
Comment (johanstokking) [inline doc/content/integrations/node-red/receive/_index.md:27]: I find this all quite a lot of text (not just this paragraph but the guide in general). If you can shorten this a bit, without losing clarity, then that would be great, otherwise it's also fine with me.

File: doc/content/integrations/node-red/receive/_index.md [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+To listen to uplink messages coming from end devices, subscribe to `v3/{application_id}/devices/{device_id}/up`, where `{application_id}` and `{device_id}` are the application and device you wish to connect to.
```
Comment (johanstokking) [inline doc/content/integrations/node-red/receive/_index.md:29]: Refer again to the `/integrations/mqtt` page here

Note that in a multi-tenant TTES setup (not this repo), this looks a bit different. Please file a follow up PR there

File: doc/content/integrations/node-red/send/_index.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+```json
+{
+  "downlinks": [{
+    "f_port": 15,
+    "frm_payload": "vu8=",
+    "priority": "NORMAL"
+  }]
+}
+```
+
```
Comment (johanstokking) [inline doc/content/integrations/node-red/send/_index.md:30]: I would split this in two pieces. It's pretty hard to get the right base64 payload manually in here.

The two pieces would be the **inject** node that creates a buffer, and drafting the JSON payload to be published.

Here's an example:

```json
[{"id":"9602047.883b9f8","type":"inject","z":"787d143a.05e53c","name":"","topic":"","payload":"[0, 42, 255, 0]","payloadType":"bin","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":190,"y":380,"wires":[["b691b63f.189828"]]},{"id":"5ec92ae4.08b554","type":"debug","z":"787d143a.05e53c","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","x":690,"y":380,"wires":[]},{"id":"b691b63f.189828","type":"function","z":"787d143a.05e53c","name":"convert to downlink message","func":"return {\n  payload: {\n    downlinks: [\n        {\n            f_port: 15,\n            frm_payload: msg.payload.toString('base64'),\n            priority: \"NORMAL\"\n        }\n    ]\n  }\n}","outputs":1,"noerr":0,"x":430,"y":380,"wires":[["5ec92ae4.08b554"]]}]
```

(use menu -> **Import** to import this)

You basically get this:

Screenshot 2020-06-03 at 11 57 59

File: doc/content/integrations/node-red/receive/_index.md (hunk 19 above)
Comment (laurensslats) [reply doc/content/integrations/node-red/receive/_index.md:13]: Second that. It's not clear where to look on the printscreens. It takes too much time to look for which buttons to press and where the add the content.

File: doc/content/integrations/node-red/send/_index.md [hunk 24]
```diff
@@ -4,32 +4,50 @@ description: ""
...
+4. Choose a **QoS** from listed options and state whether you want the MQTT Server to retain messages. 
 
 {{< figure src="mqtt_out_node_properties.png" alt="mqtt out node properties" >}}
 
-Further, find and place the **inject** node on the dashboard. This node injects a message into a flow and can be triggered manually, but it is also possible to define regular intervals between the automatic injections. Double-click on the **inject** node to configure its properties. The **Payload** field can have different formats - in this example, JSON is used. Expand the **Payload** field and enter this:
+## Configure Inject Node
+
+1. Place the **inject** node on the dashboard. Double-click on the node to configure its properties. 
+
+2. Choose **buffer** under **Payload** and enter the payload you wish to send.
```
Comment (johanstokking) [inline doc/content/integrations/node-red/send/_index.md:27]: So this is in array form, right, so `[0, 1, 2, 4]` for bytes `01020304` (as hex)

File: doc/content/integrations/node-red/send/_index.md [hunk 25]
```diff
@@ -4,32 +4,50 @@ description: ""
...
+
+```bash
+return {
+  "payload": {
+    "downlinks": [{
+      "f_port": 15,
+      "frm_payload": msg.payload.toString('base64'),
+      "priority": "NORMAL"
+    }]
+  }
```
Comment (johanstokking) [inline doc/content/integrations/node-red/send/_index.md:47]: nit; inconsistent quotes (`"` vs `'`)

File: doc/content/integrations/node-red/_index.md [hunk 26]
```diff
@@ -0,0 +1,10 @@
+---
+title: "Node-RED"
+description: ""
+---
+
+[Node-RED](https://nodered.org/) is a free, JavaScript-based development tool for visual programming, developed to ease the process of wiring together hardware devices, APIs and online services.
+
+The MQTT server that is exposed by {{% tts %}} can be connected to Node-RED. This integration allows setting up a Node-RED flow that listens to events and uplink messages, and it is also possible to schedule downlink messages.
+
+<!--more-->
```
Comment (johanstokking) [inline doc/content/integrations/node-red/_index.md:10]: Bonus points for adding a (cropped) screenshot here with a demo setup

File: doc/content/integrations/node-red/receive/_index.md [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+
+>Note: in this example, TLS-secured connection is to be established, so the **Port** value is set to 8883. In this case you should also check the **Enable secure (SSL/TLS) connection** box.
+
+{{< figure src="mqtt_in_node_connection.png" alt="Configuring MQTT Server connection information" >}}
+
+4. In the **Security** tab, enter the **Username** and **Password** according to the values in {{% tts %}} Console.
+
+{{< figure src="mqtt_in_node_security.png" alt="Configuring MQTT Server credentials" >}}
+
+5. Go back to **Properties** and set the **Topic** value to `#` (to listen to all topics).
```
Comment (johanstokking) [inline doc/content/integrations/node-red/receive/_index.md:25]: listen -> subscribe

File: doc/content/integrations/node-red/_index.md [hunk 28]
```diff
@@ -0,0 +1,12 @@
+---
+title: "Node-RED"
+description: ""
+---
+
+[Node-RED](https://nodered.org/) is a free, JavaScript-based development tool for visual programming, developed to ease the process of wiring together hardware devices, APIs and online services.
+
+The MQTT server that is exposed by {{% tts %}} can be connected to Node-RED. This integration allows setting up a Node-RED flow that listens to events and uplink messages, and it is also possible to schedule downlink messages.
+
+{{< figure src="demo_setup.png" alt="Demo setup" >}}
```
Comment (htdvisser) [inline doc/content/integrations/node-red/_index.md:10]: Not sure we should put the image in the summary

Comment (nejraselimovic) [reply doc/content/integrations/node-red/_index.md:10]: @johanstokking mentioned adding a cropped screenshot with a demo setup here. Should I remove this?

Comment (htdvisser) [reply doc/content/integrations/node-red/_index.md:10]: I don't know if @johanstokking meant in the summary or on the overview page

Comment (johanstokking) [reply doc/content/integrations/node-red/_index.md:10]: I meant in the overview page, but under the more-tag, so it's not in the summary but in the overview page

File: doc/content/integrations/node-red/receive/_index.md [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+title: "Receive Events and Messages"
+description: ""
+weight: 2
+---
+
+This section follows the process of setting up a flow which subscribes and listens to the events and messages that are being published by the MQTT Server. 
+
+## Configure MQTT In Node
+
+1. Place the **mqtt in** node on the dashboard. Double-click on the node to configure its properties.
```
Comment (benolayinka) [inline doc/content/integrations/node-red/receive/_index.md:11]: We don't use numbered instructions anywhere else in the documentation and it goes against our [style guidelines](https://github.com/TheThingsNetwork/lorawan-stack/blob/develop/DEVELOPMENT.md#documentation-site). Keep the short paragraphs, but please remove the numbers.
